docs_nnx/guides/pytree.md

```diff
@@ -8,60 +8,159 @@ jupytext:
+
+rngs = nnx.Rngs(0)
+weights = Linear(2, 3, rngs=rngs)
+
+@jax.jit
+def forward(weights, x):
+  return x @ weights.kernel
+
+y = forward(weights, x=rngs.uniform((5, 2)))
+print(f"{y.shape = }")
+```
+
+Here `weights`, of type `Linear`, was able to be passed directly to the `jit`-ed function `forward`. Throughout the rest of this guide we will try to answer the questions:
+1. What are pytrees? 
+2. How does NNX implement pytrees?
+
++++
+
+## Pytrees 101
+JAX pytrees are tree structures made of python objects that can be recursively traversed in order to collect an ordered list of leaves and a definition of the tree structure, this is done via the `jax.tree.flatten` function. Most common pytrees are native python containers like `list`, `dict`, and `tuple`, but interestingly it also include `None`. The example bellow shows how to collect all the integer leaves from a nested structure using `flatten`:
+
+```{code-cell} ipython3
+pytree = [
+  {'a': 1},
+  {
+    'b': 2,
+    'c': (3, 4),
+    'd': None,
+  }
+]
+
+leaves, treedef = jax.tree.flatten(pytree)
+print(f"leaves = {leaves}")
+print(f"treedef = {treedef}")
+```
+
+Note that `None` is not a leaf because its defined as a pytree with no children. The main purpose of being able to flatten, apart from collecting the leaves, is being able reconstruct the pytree structure from the tree definition from any sequence of leaves of the same length via the `jax.tree.unflatten` function:
+
+```{code-cell} ipython3
+new_leaves = [x * 10 for x in leaves]
+new_pytree = jax.tree.unflatten(treedef, new_leaves)
+
+print(f"old pytree = {pytree}")
+print(f"new pytree = {new_pytree}")
 ```
 
-## Pytree
+### Custom Pytrees
+JAX allows us to register custom pytree node type by using the `jax.tree_util.register_pytree_node` utility. For any type we are able to define a flatten that decomposes the object into a a sequence of nodes / children and a static (hashable) structure, and a unflatten function which takes the sequence of nodes and the static structure and creates a new instance. In the following example we create a simple type `Foo` with the attributes `a`, `b`, and `c`, and define `a` and `b` as nodes and `c` as static.
+
+```{code-cell} ipython3
+class Foo:
+  def __init__(self):
+    self.a = 1
+    self.b = 2
+    self.c = "hi"
+
+def flatten_foo(foo: Foo):
+  nodes = [foo.a, foo.b]  # sequence of nodes
+  static = (foo.c,) # hashable & equatable structure
+  return nodes, static
+
+def unflatten_foo(static, nodes):
+  foo = object.__new__(Foo)  # create uninitialized instance
+  foo.a = nodes[0]
+  foo.b = nodes[1]
+  foo.c = static[0]
+  return foo
+
+jax.tree_util.register_pytree_node(Foo, flatten_foo, unflatten_foo)
+
+foo = Foo()
+leaves, treedef = jax.tree.flatten(foo)
+print(f"leaves = {leaves}")
+print(f"treedef = {treedef}")
+```
+
+Notice that `'hi'` does not appear in the leaves because `c` is defined as static, but you can see it as part of the `PyTreeDef` structure.
+
```

Comment:
Before diving into details of pytree, I kinda want there to be a line explaining *why* we care about pytrees in JAX ecosystem. 
Even a simple mention like saying "JAX uses pytrees to contain and track array data, to optimize its compilations" would do.